rs/boundary_node/salt_sharing/canister/canister.rs

```diff
@@ -0,0 +1,75 @@
+use crate::storage::{Timestamp, LAST_SALT_NS, SALT, SALT_SIZE};
+use crate::time::delay_till_next_month;
+use getrandom::getrandom;
+use ic_cdk::api::time;
+use ic_cdk_macros::{init, post_upgrade, query};
+use ic_cdk_timers::set_timer;
+use salt_api::{GetSaltResponse, InitArg, SaltGenerationStrategy, SaltResponse};
+
+// Sets an execution timer (delayed future task) and returns immediately.
+fn reschedule_salt_generation(strategy: SaltGenerationStrategy) {
+    match strategy {
+        SaltGenerationStrategy::StartOfMonth => {
+            let delay = delay_till_next_month(time());
+            set_timer(delay, || {
+                regenerate_salt();
+                // Function is called recursively to schedule next execution
+                reschedule_salt_generation(strategy);
+            });
+        }
+    }
+}
+
+// Run when the canister is first installed
+#[init]
+fn init(init_arg: InitArg) {
+    // Generate salt on the very first init or based on the provided argument.
+    if !is_salt_init() || init_arg.regenerate_now {
+        regenerate_salt();
+    }
+    // Start salt generation schedule based on the argument.
+    if let Some(strategy) = init_arg.salt_generation_strategy {
```

Comment:
If we set the strategy again, will it cancel the old timer or will it then have two timers: then one from before and the new one?

Reply:
When the canister is upgraded all [existing timers will be stopped](https://internetcomputer.org/docs/current/developer-docs/backend/rust/timers#implementing-canister-upgrade). and should be restarted manually. That's why there is a all to init in the `post_upgrade`.

Reply:
You could also consider persisting the config in stable memory so that you don't need to repeat upon every upgrade unless there's a change to the config.

Reply:
I don't think I understand what You mean here. You refer to `SaltGenerationStrategy` as "config", right?  We anyway need to restart the `schedule_salt_generation()` function on every upgrade.

Reply:
> You refer to SaltGenerationStrategy as "config", right?

Indeed.

> We anyway need to restart the schedule_salt_generation() function on every upgrade.

That's true, but you might retrieve the configs from stable memory instead of requiring every canister upgrade to contain the (same) configs every time.